algebird-core/src/main/scala/com/twitter/algebird/maps/redblack.scala

```diff
@@ -0,0 +1,219 @@
+        case (bl, BNode(y, a, b)) => balanceDel(x, bl, rNode(y, a, b))
+        case (bl, RNode(y, BNode(z, a, b), c)) =>
+          rNode(z, bNode(x, bl, a), balanceDel(y, b, redden(c)))
+        case _ => throw new Exception(s"undefined pattern in tree pair: ($tl, $tr)")
+      }
+
+    final protected def balanceRight(x: Data[K], tl: Node[K], tr: Node[K]) =
+      (tl, tr) match {
+        case (a, RNode(y, b, c)) => rNode(x, a, bNode(y, b, c))
+        case (BNode(y, a, b), bl) => balanceDel(x, rNode(y, a, b), bl)
+        case (RNode(y, a, BNode(z, b, c)), bl) =>
+          rNode(z, balanceDel(y, redden(a), b), bNode(x, c, bl))
+        case _ => throw new Exception(s"undefined pattern in tree pair: ($tl, $tr)")
+      }
+
+    final protected def append(tl: Node[K], tr: Node[K]): Node[K] = (tl, tr) match {
+      case (LNode(), n) => n
+      case (n, LNode()) => n
+      case (RNode(x, a, b), RNode(y, c, d)) => append(b, c) match {
+        case RNode(z, bb, cc) => rNode(z, rNode(x, a, bb), rNode(y, cc, d))
+        case bc => rNode(x, a, rNode(y, bc, d))
+      }
+      case (BNode(x, a, b), BNode(y, c, d)) => append(b, c) match {
+        case RNode(z, bb, cc) => rNode(z, bNode(x, a, bb), bNode(y, cc, d))
+        case bc => balanceLeft(x, a, bNode(y, bc, d))
+      }
+      case (a, RNode(x, b, c)) => rNode(x, append(a, b), c)
+      case (RNode(x, a, b), c) => rNode(x, a, append(b, c))
+    }
+
+    // NOTE: the balancing rules for node deletion all assume that the case of deleting a key
+    // that is not in the map is addressed elsewhere.  If these balancing functions are applied
+    // to a key that isn't present, they will fail destructively and uninformatively.
+    final protected def delLeft(node: INode[K], k: K) = node.lsub match {
+      case n: INode[K] if (n.color == B) => balanceLeft(node.data, node.lsub.del(k), node.rsub)
+      case _ => rNode(node.data, node.lsub.del(k), node.rsub)
+    }
+
+    final protected def delRight(node: INode[K], k: K) = node.rsub match {
+      case n: INode[K] if (n.color == B) => balanceRight(node.data, node.lsub, node.rsub.del(k))
+      case _ => rNode(node.data, node.lsub, node.rsub.del(k))
+    }
+  }
+
+  /** Represents a leaf node in the Red Black tree system */
+  trait LNode[K] extends Node[K] {
+    final def node(k: K) = None
+
+    final def nodeMin = None
+    final def nodeMax = None
+
+    final def ins(d: Data[K]) = rNode(d, this, this)
+    final def del(k: K) = this
+  }
+
+  private object LNode {
+    def unapply[K](node: LNode[K]): Boolean = true
+  }
+
+  /** Represents an internal node (Red or Black) in the Red Black tree system */
+  trait INode[K] extends Node[K] {
+    /** The Red/Black color of this node */
+    val color: Color
+    /** Including, but not limited to, the key */
+    val data: Data[K]
+    /** The left sub-tree */
+    val lsub: Node[K]
+    /** The right sub-tree */
+    val rsub: Node[K]
+
```

Comment:
I think updating these to be `def` instead of `val` will be relatively straightforward